pyomo/core/plugins/transform/scaling.py

```diff
@@ -0,0 +1,269 @@
+        >>> print(value(model.obj))
+        101000000.0
+        >>> print(value(scaled_model.obj))
+        101.0
+
+    """
+
+    alias('core.scale_model',\
+          doc="Create a model where the variables and constraints are scaled by user-provided values.")
+
+    def __init__(self, **kwds):
+        kwds['name'] = "scale_model"
+        self._scaling_method = kwds.pop('scaling_method', 'user')
+        super(ScaleModel, self).__init__(**kwds)
+
+    def _create_using(self, original_model, **kwds):
+        scaled_model = original_model.clone()
+        self._apply_to(scaled_model, **kwds)
+        return scaled_model
+    
+    def _get_float_scaling_factor(self, instance, component_data):
+        scaling_factor = None
+        if component_data in instance.scaling_factor:
+            scaling_factor = instance.scaling_factor[component_data]
+        elif component_data.parent_component() in instance.scaling_factor:
+            scaling_factor = instance.scaling_factor[component_data.parent_component()]
+            
+        if scaling_factor is None:
+            return 1.0
+        
+        try:
+            scaling_factor = float(scaling_factor)
+        except ValueError:
+            raise ValueError("Suffix 'scaling_factor' has a value %s for component %s that cannot be converted to a float. "
+                             "Floating point values are required for this suffix in the ScaleModel transformation."
+                             % (scaling_factor, component_data))
+        return scaling_factor
+
+    def _apply_to(self, model, **kwds):
+        # need to have a map of unscaled components to scaling factors
+        # and a map of scaled components to unscaled components
+        component_scaling_factor_map = ComponentMap()
+        scaled_to_original_map = None
+        
+        # if the scaling_method is 'user', get the scaling parameters from the suffixes
+        if self._scaling_method == 'user':
+            # perform some checks to make sure we have the necessary suffixes
+            if type(model.component('scaling_factor')) is not Suffix:
+                raise ValueError("ScaleModel transformation called with scaling_method='user'"
+                                 ", but cannot find the suffix 'scaling_factor' on the model")
+
+            # get the scaling factors
+            for c in model.component_data_objects(ctype=(Var,Constraint,Objective), descend_into=True):
+                component_scaling_factor_map[c] = self._get_float_scaling_factor(model, c)
+
+        else:
+            raise ValueError("ScaleModel transformation: unknown scaling_method found"
+                             "-- supported values: 'user' ")
+
+        # scale the variable bounds and values and build the
+        # variable substitution map for scaling vars in constraints
+        variable_substitution_map = ComponentMap()
+        for variable in [var for var in model.component_objects(ctype=Var, descend_into=True)]:
```

Comment:
This is weird syntax.  Why the intermediate list?  If you are worried about the tree changing while you iterate over the Vars, I would prefer:
```
var_objects = list( model.component_objects(Var, descend_into=True) )
for var in var_objects:
```